Add static invoice creation utils to `ChannelManager`

valentinewallace opened the pull request:
1. Add static invoice creation utilities as part of supporting the async payments BOLTs spec https://github.com/lightning/bolts/pull/1149. 

2. Take this opportunity to more easily test some code added in https://github.com/lightningdevkit/rust-lightning/pull/3140 that went untested at the time. This is the bulk of the diff. 

3. Address a piece of feedback from #3140 regarding `InvoiceRequest`s being unavailable when the time comes to send the async payment, cc https://github.com/lightningdevkit/rust-lightning/pull/3140#discussion_r1759425911

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10055,14 +10055,20 @@ where
-		let max_cltv_expiry = self.best_block.read().unwrap().height + CLTV_FAR_FAR_AWAY
-			+ LATENCY_GRACE_PERIOD_BLOCKS;
+
+		const SECONDS_PER_BLOCK: u32 = 10 * 60;
+		let relative_expiry_blocks = relative_expiry_seconds / SECONDS_PER_BLOCK;
+		let max_cltv_expiry = core::cmp::max(relative_expiry_blocks, CLTV_FAR_FAR_AWAY)
```
Note for reviewers: I'm not sure this is robust enough with block times being somewhat unreliable. Maybe it should be buffered?

vincenzopalazzo commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -134,13 +134,16 @@ pub(crate) enum PendingOutboundPayment {
 }
 
+#[derive(Clone)]
 pub(crate) struct RetryableInvoiceRequest {
 	pub(crate) invoice_request: InvoiceRequest,
 	pub(crate) nonce: Nonce,
+	pub(super) needs_retry: bool,
 }
```
IIRC there is nothing in the offers, just at the invoice request level.

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -104,6 +104,9 @@ impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, ES: Deref, S: Deref, SP: Size
 		// recipient's node_id.
 		const MIN_PEER_CHANNELS: usize = 3;
 
+		const DEFAULT_AMT_MSAT: u64 = 100_000_000;
+		let amount_msats = amount_msats.unwrap_or(DEFAULT_AMT_MSAT);
```
It was chosen pretty arbitrarily lol. Input welcome on it. I'm guessing async receivers are mostly doing p2p payments and this amount is about $100 USD right now. I think the main constraint is that we don't want to exceed the recipient's channel capacity, and they may only have 1 channel with their LSP. Another option would be to not filter channels by capacity/{min,max}_htlc if no amount is provided.

valentinewallace replied:
Thinking on this more, I think the amount as-is is too high. Will explore either lower amounts or skipping some of the current filters.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9549,6 +9552,86 @@ where
+		let payment_context = PaymentContext::AsyncBolt12Offer(
+			AsyncBolt12OfferContext { offer_id: offer.id(), offer_nonce }
+		);
+		let amount_msat = offer.amount().and_then(|amount| {
+			match amount {
+				Amount::Bitcoin { amount_msats } => Some(amount_msats),
+				Amount::Currency { .. } => None
+			}
+		});
```
`quantity_max` is currently allowed. I'm not sure I follow your question though:

> If so, is the amount checked by them in some way?

Who is "them"? It looks like the payer checks the amount/quantity when sending an invreq but not sure I'm following.

jkczyz replied:
Does the recipient check that the amount is sufficient for the requested quantity before releasing the preimage?

valentinewallace replied:
Yeah seems they will have to, if that isn't done already. Currently we don't support receiving to the static invoices created by these utils (just creating them), so this is definitely something to consider in the follow-up.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9549,6 +9552,86 @@ where
+			node_id, expanded_key, nonce, secp_ctx
+		).chain_hash(self.chain_hash);
+
+		for path in message_paths_to_always_online_node {
+			builder = builder.path(path);
+		}
```
> If the recipient is online, does the InvoiceRequest get forwarded along?

Yep!

> how does the recipient authenticate it?

Since the recipient issued the offer themselves, they authenticate it the same way as always-online recipients, i.e. via `verify_using_recipient_data`, if I'm understanding you.

jkczyz replied:
Right, so I guess that means the recipient must construct these paths since it needs to include the `Nonce` used to derive the signing keys.

jkczyz replied:
Hmm... but the function is creating and returning the `Nonce`. I think we'll need to pass it in?

jkczyz replied:
To be clear, we don't include metadata in the offer when we have blinded paths. Can't recall if we do something different when constructing an offer for use with async payments.

jkczyz replied:
I brought this up during review club. Since payment path contains the offer nonce -- and the sender will include the invoice request with the payment -- we should be able to verify the invoice request is authentic with this nonce. But ISTM we still want the same nonce in the offer's message paths for the normal case when the recipient is online.

valentinewallace replied:
Ah I misunderstood at first, this is a good point.... I'm not sure yet what a good solution would be.  

For context, `message_paths_to_always_online_node` terminate at the always-online node. Currently in LDK the recipient expects to receive the invreq either over a blinded path from the offer or to their node id but with `offer_metadata` set (IIUC), but neither of those are supported in this PR atm. 

For more context, I think the always-online node is expected to be a direct peer of the async receiver, otherwise an onion message round-trip would be incurred to check whether the recipient is online when the always-online node receives the invreq. 

So it seems like our two options here are to either set the offer metadata for async receive offers (which may be a privacy leak), or for the recipient to provide the always-online node with a one-hop blinded path to send the invreq over, separately from their static invoice. Either of these options should allow the recipient to verify the invreq if they happen to be online at the time. 

Any thoughts on this?

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9549,6 +9552,86 @@ where
+	/// Create an offer for receiving async payments as an often-offline recipient.
+	///
+	/// Because we may be offline when the payer attempts to request an invoice, you MUST:
+	/// 1. Provide at least 1 [`BlindedPath`] for onion messages terminating at an always-online node
```
Note to self: reference `BlindedMessagePath` here instead

valentinewallace commented on lightning/src/blinded_path/payment.rs:
```diff
@@ -351,6 +357,22 @@ pub struct Bolt12OfferContext {
+///
+/// [`Offer`]: crate::offers::offer::Offer
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AsyncBolt12OfferContext {
```
Tracking this on #2298. I'm inclined to build on Jeff's solution to #3427 to make sure we align with the rest of the offers code unless there are objections.

valentinewallace commented on lightning/src/blinded_path/payment.rs:
```diff
@@ -351,6 +357,22 @@ pub struct Bolt12OfferContext {
+	/// The identifier of the [`Offer`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	pub offer_id: OfferId,
```
This is used to fill in the `Bolt12OfferContext::offer_id` field. I ended up opening the follow-up if you want to check out more context: https://github.com/lightningdevkit/rust-lightning/pull/3440

valentinewallace replied:
This `offer_id` is in the static invoice for its blinded payment path contexts, not the offer

valentinewallace replied:
Ah, that's true, it looks like we actually already re-calculate the `offer_id` when calling `InvoiceRequest::verify_using_recipient_data`. I think it might not be exposed currently but should be able to add an accessor

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9549,6 +9557,86 @@ where
+	/// created via [`Self::create_async_receive_offer_builder`]. If `relative_expiry` is unset, the
+	/// invoice's expiry will default to [`STATIC_INVOICE_DEFAULT_RELATIVE_EXPIRY`].
+	#[cfg(async_payments)]
+	pub fn create_static_invoice_builder_for_async_receive_offer<'a>(
```
Should we drop the `_for_async_receive_offer` suffix? Will there be another utility that has a similar name?

valentinewallace replied:
> Will there be another utility that has a similar name?

Nah, I just wanted to make it clear that you can't create these static invoices from regular offers. I can drop it though if you prefer.

jkczyz commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -1416,6 +1424,209 @@ fn custom_tlvs_to_blinded_path() {
+	let offer = offer_builder.build().unwrap();
+	let amt_msat = 5000;
+	let payment_id = PaymentId([1; 32]);
+	// Set the random bytes so we can predict the offer outbound payment context nonce.
+	*nodes[0].keys_manager.override_random_bytes.lock().unwrap() = Some([42; 32]);
```
Shouldn't these tests go in `offers_tests.rs` and use `extract_offer_nonce`?

valentinewallace replied:
We can't use `extract_offer_nonce` because the nonce we're trying to extract comes from the reply path to the invreq, not the invreq itself. But I switched away from manually constructing the hmac, see the latest fixup!

jkczyz reviewed (commented):
Fix-ups LGTM. Just comments on the tests.

jkczyz commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -1497,25 +1512,25 @@ fn ignore_unexpected_static_invoice() {
+	let invreq_om = nodes[0].onion_messenger.next_onion_message_for_peer(nodes[1].node.get_our_node_id()).unwrap();
+	let invreq_reply_path = extract_invoice_request_reply_path(&nodes[1], &invreq_om);
 
 	// Create a static invoice with the same payment_id but corresponding to a different offer.
```
IIUC, the static invoice doesn't have a payment id. Rather the reply path that it is sent over does.

jkczyz commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -1416,6 +1444,209 @@ fn custom_tlvs_to_blinded_path() {
+	let invreq_om = nodes[0].onion_messenger.next_onion_message_for_peer(nodes[1].node.get_our_node_id()).unwrap();
+	let invreq_reply_path = extract_invoice_request_reply_path(&nodes[1], &invreq_om);
+
+	// Create a static invoice with the same payment_id but corresponding to a different offer.
```
s/Create/Send

valentinewallace replied:
Here we're creating the invoice, though?

jkczyz replied:
Ah, right. Sorry, I was looking at a different place which you already fixed.

jkczyz reviewed (commented):
LGTM. Just some minor comments to address.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9549,6 +9557,86 @@ where
+	/// Because we may be offline when the payer attempts to request an invoice, you MUST:
+	/// 1. Provide at least 1 [`BlindedMessagePath`] terminating at an always-online node that will
+	///    serve the [`StaticInvoice`] created from this offer on our behalf.
+	/// 2. Use [`Self::create_static_invoice_builder_for_async_receive_offer`] to create a
```
Needs updating

jkczyz reviewed (commented):
Feel free to squash after fixing the doc comment

valentinewallace commented on lightning/src/blinded_path/message.rs:
```diff
@@ -402,6 +403,28 @@ pub enum AsyncPaymentsContext {
+		/// one of our offers.
+		///
+		/// [`Offer`]: crate::offers::offer::Offer
+		offer_id: OfferId,
```
This is basically just used as an input into the HMAC field 10 lines below, so it helps authenticate this context. But maybe the nonce is sufficient as a unique input? I was following the pattern of `hmac_for_payment_id` used for regular bolt 12 invoices but using an `offer_id` instead.

valentinewallace replied:
Yeah, I didn't really have a sense of whether nonce + hmac was sufficient since we haven't used only those things previously, but since it is I went ahead and removed it. Will track expiring these paths on the async payments issue.

valentinewallace replied:
To be clear, in this scenario the node trying to ask if we're online has our node id. Another way for nodes to trivially ask if we're online is just by attempting to connect to us as a peer -- but presumably in this case the attacker is prevented from doing it that way by not knowing our socket address?

valentinewallace replied:
Ok I started doing this but tbh I'd prefer to save it for follow-up if it's all the same to you. 

> No, I'm thinking of a mobile node that's only connect to an LSP and for which we only have a blinded path to.

Ok so for context I think that means the attacker would be reusing the `blinded_node_id`s to construct their path. What I'm not following is that in order to blind the TLVs of their choosing, I thought they would need the shared secret corresponding to the node's unblinded pubkey and the blinding point?

jkczyz replied:
> Arguably it would be reasonable to reply to _every_ inbound `HeldHtlcAvailable` message by just asking that the HTLC be released. It would imply someone can trivially ask if/when you're online which isn't ideal, so we should have _something_ here (a nonce/hmac pair seems to suffice), but ultimately we may get a bunch of these for any given offer, so its not like we can track them and remove offers after they're paid. We might want a timeout here so that we can stop replying to old ones that have expired, I guess?

IIUC, this context is in the static invoice's message paths. So couldn't an adversary simply request an invoice, get back the static invoice, and save the path to use whenever it wants? We could have it expire with the static invoice, but there doesn't seem to be anything that would stop the adversary from doing the same thing with a new static invoice. Am I missing something?

valentinewallace replied:
You're not missing anything IIUC. The only thing that limits the attack you're describing is also expiring the offer/its paths.

jkczyz replied:
But isn't this a *message* path from the static invoice? Would we need to add an expiration to this new context in order to expire it? For payment paths we'd use the payment secret, IIUC.

valentinewallace replied:
Discussed offline; it is the case that if a recipient publishes an offer that never expires, then anyone can pretty much always check if they're online under this protocol as long as the offer's message paths remain valid. Probably worth noting in documentation in the future.

jkczyz replied:
Chatted offline. Just was trying to point out that if you _always_ want to be paid by _anyone_, then anyone can check if you are online because presumably you will always have a fresh offer with unexpired paths published.

valentinewallace commented:
Removed the `offer_id` field from the payment context, squashed with this diff:
```diff
diff --git a/lightning/src/blinded_path/payment.rs b/lightning/src/blinded_path/payment.rs
index 70ac9823e..d72c13353 100644
--- a/lightning/src/blinded_path/payment.rs
+++ b/lightning/src/blinded_path/payment.rs
@@ -362,10 +362,6 @@ pub struct Bolt12OfferContext {
 /// [`Offer`]: crate::offers::offer::Offer
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub struct AsyncBolt12OfferContext {
-       /// The identifier of the [`Offer`].
-       ///
-       /// [`Offer`]: crate::offers::offer::Offer
-       pub offer_id: OfferId,
        /// The [`Nonce`] used to verify that an inbound [`InvoiceRequest`] corresponds to this static
        /// invoice's offer.
        ///
@@ -650,8 +646,7 @@ impl_writeable_tlv_based!(Bolt12OfferContext, {
 });

 impl_writeable_tlv_based!(AsyncBolt12OfferContext, {
-       (0, offer_id, required),
-       (2, offer_nonce, required),
+       (0, offer_nonce, required),
 });

 impl_writeable_tlv_based!(Bolt12RefundContext, {});
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index ba21b87ac..ed31e00c6 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -9602,7 +9602,7 @@ where
                let secp_ctx = &self.secp_ctx;

                let payment_context = PaymentContext::AsyncBolt12Offer(
-                       AsyncBolt12OfferContext { offer_id: offer.id(), offer_nonce }
+                       AsyncBolt12OfferContext { offer_nonce }
                );
                let amount_msat = offer.amount().and_then(|amount| {
                        match amount {

valentinewallace commented on lightning/src/blinded_path/message.rs:
```diff
@@ -402,6 +402,22 @@ pub enum AsyncPaymentsContext {
+		///
+		/// [`HeldHtlcAvailable`]: crate::onion_message::async_payments::HeldHtlcAvailable
+		nonce: Nonce,
+		/// Authentication code for the [`OfferId`].
+		///
+		/// Prevents the recipient from being able to deanonymize us by creating a blinded path to us
+		/// containing the expected [`OfferId`].
```
Oops, will update these docs on the next push.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -46,6 +46,11 @@ const ASYNC_PAYMENT_ID_HMAC_INPUT: &[u8; 16] = &[6; 16];
+// HMAC input used in `AsyncPaymentsContext::InboundPayment` to authenticate inbound
+// held_htlc_available onion messages.
+#[cfg(async_payments)]
+const ASYNC_PAYMENTS_HELD_HTLC_HMAC_INPUT: &[u8; 16] = &[8; 16];
```
Will need to be updated after rebase.

valentinewallace commented:
Was testing the follow-up and realized we weren't allowing MPP on the static invoices:
```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 112b2a0fa..ef118b99b 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -10068,7 +10068,7 @@ where
                StaticInvoiceBuilder::for_offer_using_derived_keys(
                        offer, payment_paths, async_receive_message_paths, created_at, expanded_key,
                        offer_nonce, secp_ctx
-               )
+               ).map(|inv| inv.allow_mpp())
        }

        /// Pays for an [`Offer`] using the given parameters by creating an [`InvoiceRequest`] and

jkczyz commented:
> Was testing the follow-up and realized we weren't allowing MPP on the static invoices:

Should we leave this to the user? Otherwise, there's no way for them to disable this. Then again, we don't provide them a way to disable it for normal invoices.

valentinewallace commented:
> > Was testing the follow-up and realized we weren't allowing MPP on the static invoices:
> 
> Should we leave this to the user? Otherwise, there's no way for them to disable this. Then again, we don't provide them a way to disable it for normal invoices.

Fair point... I definitely think it's a better UX to enable MPP by default. IIRC, the only reason we've ran into for disabling it has been phantom payments, which aren't currently supported. I could document a follow-up to add an API for disabling it, possibly.

valentinewallace commented:
Went ahead and moved the tests added here (and a couple pre-existing tests) into a dedicated async payments tests file. A good amount more tests are getting added in the follow-up so figured this change makes sense.

jkczyz reviewed (commented):
LGTM. Feel free to squash IMO.

valentinewallace commented:
Err, also wasn't setting the invoice's relative expiry outside of the blinded paths.
```diff
diff --git a/lightning/src/ln/async_payments_tests.rs b/lightning/src/ln/async_payments_tests.rs
index 41c1ec33b..424d76da6 100644
--- a/lightning/src/ln/async_payments_tests.rs
+++ b/lightning/src/ln/async_payments_tests.rs
@@ -35,6 +35,7 @@ use crate::util::config::UserConfig;
 use bitcoin::secp256k1::Secp256k1;

 use core::convert::Infallible;
+use core::time::Duration;

 #[test]
 fn blinded_keysend() {
@@ -514,12 +515,15 @@ fn pays_static_invoice() {
        let offer = offer_builder.build().unwrap();
        let amt_msat = 5000;
        let payment_id = PaymentId([1; 32]);
+       let relative_expiry = Duration::from_secs(1000);
        let static_invoice = nodes[2]
                .node
-               .create_static_invoice_builder(&offer, offer_nonce, None)
+               .create_static_invoice_builder(&offer, offer_nonce, Some(relative_expiry))
                .unwrap()
                .build_and_sign(&secp_ctx)
                .unwrap();
+       assert!(static_invoice.invoice_features().supports_basic_mpp());
+       assert_eq!(static_invoice.relative_expiry(), relative_expiry);

        nodes[0]
                .node
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index ef118b99b..755198a8e 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -10068,7 +10068,7 @@ where
                StaticInvoiceBuilder::for_offer_using_derived_keys(
                        offer, payment_paths, async_receive_message_paths, created_at, expanded_key,
                        offer_nonce, secp_ctx
-               ).map(|inv| inv.allow_mpp())
+               ).map(|inv| inv.allow_mpp().relative_expiry(relative_expiry_secs))
        }

        /// Pays for an [`Offer`] using the given parameters by creating an [`InvoiceRequest`] and

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -102,8 +102,8 @@ pub struct StaticInvoiceBuilder<'a> {
 	///
-	/// Unless [`StaticInvoiceBuilder::relative_expiry`] is set, the invoice will expire 24 hours
-	/// after `created_at`.
+	/// The invoice's expiration will default to [`DEFAULT_RELATIVE_EXPIRY`] after `created_at` unless
+	/// overridden by [`StaticInvoiceBuilder::relative_expiry`].
```
IIUC, this default is not defined by the spec (which uses 2 hours for any invoice), and thus if not set `StaticInvoice:: relative_expiry` would give 24 hours whereas another implementation would assume it is 2 hours.

valentinewallace replied:
Good point, thanks. I'll need to update the spec for that (added to the tracking issue for now). Currently `relative_expiry()` gives 2 weeks if not set, though we may want to revisit that.

jkczyz replied:
Do we still need to explicitly set it in the `ChannelManager` utility then as was added in the recent push? I guess it doesn't matter much since size isn't a constraint, but just wondering if a test caught this in a follow-up?

valentinewallace replied:
Yeah, caught this in testing: https://github.com/lightningdevkit/rust-lightning/pull/3517/commits/7d239c4f2507bc4ec81ddede2a1e81d51472a397. Fair that we could avoid setting it if it's `None` or the default, though, let me know if you'd prefer that. Since size isn't a constraint as you mention, it simplifies the code a bit to always set it so I'm neutral.

jkczyz replied:
Might as well have it be explicit.
